lms/services/application_instance.py

```diff
@@ -54,6 +58,40 @@ def get_by_consumer_key(self, consumer_key) -> ApplicationInstance:
         except NoResultFound as err:
             raise ApplicationInstanceNotFound() from err
 
+    def build_from_lms_url(  # pylint:disable=too-many-arguments
+        self,
+        lms_url,
+        email,
+        developer_key,
+        developer_secret,
+        settings,
+    ):
+        """Instantiate ApplicationInstance with lms_url."""
+        encrypted_secret = developer_secret
+        aes_iv = None
+
+        if developer_secret and developer_key:
+            aes_iv = self._aes_service.build_iv()
+            encrypted_secret = self._aes_service.encrypt(aes_iv, developer_secret)
+
+        application_instance = ApplicationInstance(
+            consumer_key="Hypothesis" + secrets.token_hex(16),
+            shared_secret=secrets.token_hex(32),
+            lms_url=lms_url,
+            requesters_email=email,
+            developer_key=developer_key,
+            developer_secret=encrypted_secret,
+            aes_cipher_iv=aes_iv,
+            created=datetime.utcnow(),
+            settings=settings,
+        )
+        self._db.add(application_instance)
+        return application_instance
+
 
 def factory(_context, request):
-    return ApplicationInstanceService(request.db, request)
+    return ApplicationInstanceService(
+        request.db,
+        request,
+        request.find_service(AESService),
```

Comment:
It's happening again! I think it's a bit odd to pass a mixture of things from the request, and the request. Does this still work if you pass the `lti_user` as I think that's all the request is used for internally?

Reply:
No, unfortunately there's a circular dependency between lti_user <-> application instance if we bring the LTIUser into the factory.

I think the right fix is a refactor/merge of security.Identity/LTIUser but I rather do that once we have the LTI1.3 stuff in so it's clearer it will work for both.